JVM seems to release versions too soon

Not sure if this behavior is only related to JVM or if other platforms have similar issues, but I have seen it appear the most on JVM.

I can reproduce after a few changes every time on my demo app for Sync.

The relevant code is:

```
    public fun adjust(change: Int) {
        // Use an application wide dispatcher
        println("Adjust $change")
        CoroutineScope(Dispatchers.Default).launch {
            realm.write {
                println("Start change on ${this.version()}")
                println("Find ${counterObj.version()}")
                findLatest(counterObj)?.run {
                    list.add(change)
                }
            }
        }
    }
```

It crashes pretty quickly with the below trace, indicating that the Realm that should have been held by `counterObj` has been closed.

```
Start change on VersionId(version=16)
Exception in thread "DefaultDispatcher-worker-3 @coroutine#1078" java.lang.IllegalStateException: Realm has been closed and is no longer accessible: ./default.realm
	at io.realm.internal.RealmReference.version(RealmReference.kt:53)
	at io.realm.internal.RealmStateHolder$DefaultImpls.version(RealmState.kt:50)
	at io.realm.internal.RealmObjectInternal$DefaultImpls.version(RealmObjectInternal.kt:37)
	at io.realm.kotlin.demo.model.Counter.version(Counter.kt:8)
	at io.realm.RealmObjectKt.version(RealmObject.kt:46)
	at io.realm.kotlin.demo.CounterRepository$adjust$1$1.invoke(CounterRepository.kt:58)
	at io.realm.kotlin.demo.CounterRepository$adjust$1$1.invoke(CounterRepository.kt:56)
	at io.realm.internal.SuspendableWriter$write$2.invokeSuspend(SuspendableWriter.kt:67)
	(Coroutine boundary)
	at io.realm.internal.RealmImpl.write(RealmImpl.kt:96)
	at io.realm.kotlin.demo.CounterRepository$adjust$1.invokeSuspend(CounterRepository.kt:56)
	(Coroutine creation stacktrace)
	at kotlin.coroutines.intrinsics.IntrinsicsKt__IntrinsicsJvmKt.createCoroutineUnintercepted(IntrinsicsJvm.kt:122)
	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:30)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:47)
	at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source)
	at io.realm.kotlin.demo.CounterRepository.adjust(CounterRepository.kt:55)
	at io.realm.kotlin.demo.MainKt$main$1$2$1$1$1$1.invoke(Main.kt:36)
	at io.realm.kotlin.demo.MainKt$main$1$2$1$1$1$1.invoke(Main.kt:35)
	at androidx.compose.foundation.ClickableKt$clickable$4$gesture$1$2.invoke-k-4lQ0M(Clickable.kt:137)
	at androidx.compose.foundation.ClickableKt$clickable$4$gesture$1$2.invoke(Clickable.kt:137)
	at androidx.compose.foundation.gestures.TapGestureDetectorKt$detectTapAndPress$2$1$1.invokeSuspend(TapGestureDetector.kt:378)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTaskKt.resume(DispatchedTask.kt:179)
	at kotlinx.coroutines.CancellableContinuationImpl.resumeWith(CancellableContinuationImpl.kt:337)
	at androidx.compose.ui.input.pointer.SuspendingPointerInputFilter$PointerEventHandlerCoroutine.offerPointerEvent(SuspendingPointerInputFilter.kt:432)
	at androidx.compose.ui.input.pointer.SuspendingPointerInputFilter.dispatchPointerEvent(SuspendingPointerInputFilter.kt:330)
	at androidx.compose.ui.input.pointer.SuspendingPointerInputFilter.onPointerEvent-H0pRuoY(SuspendingPointerInputFilter.kt:343)
	at androidx.compose.ui.input.pointer.Node.dispatchMainEventPass(HitPathTracker.kt:287)
	at androidx.compose.ui.input.pointer.Node.dispatchMainEventPass(HitPathTracker.kt:275)
	at androidx.compose.ui.input.pointer.NodeParent.dispatchMainEventPass(HitPathTracker.kt:151)
	at androidx.compose.ui.input.pointer.HitPathTracker.dispatchChanges(HitPathTracker.kt:90)
	at androidx.compose.ui.input.pointer.PointerInputEventProcessor.process-gBdvCQM(PointerInputEventProcessor.kt:77)
	at androidx.compose.ui.platform.DesktopOwner.processPointerInput-8iAsVTc$ui(DesktopOwner.desktop.kt:322)
	at androidx.compose.ui.platform.DesktopOwners.onMouseReleased(DesktopOwners.desktop.kt:177)
	at androidx.compose.ui.awt.ComposeLayer$initCanvas$2$mouseReleased$1.invoke(ComposeLayer.desktop.kt:183)
	at androidx.compose.ui.awt.ComposeLayer$initCanvas$2$mouseReleased$1.invoke(ComposeLayer.desktop.kt:182)
	at androidx.compose.ui.awt.AWTDebounceEventQueue$job$1.invokeSuspend(AWTDebounceEventQueue.desktop.kt:59)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused by: java.lang.IllegalStateException: Realm has been closed and is no longer accessible: ./default.realm
	at io.realm.internal.RealmReference.version(RealmReference.kt:53)
	at io.realm.internal.RealmStateHolder$DefaultImpls.version(RealmState.kt:50)
	at io.realm.internal.RealmObjectInternal$DefaultImpls.version(RealmObjectInternal.kt:37)
	at io.realm.kotlin.demo.model.Counter.version(Counter.kt:8)
	at io.realm.RealmObjectKt.version(RealmObject.kt:46)
	at io.realm.kotlin.demo.CounterRepository$adjust$1$1.invoke(CounterRepository.kt:58)
	at io.realm.kotlin.demo.CounterRepository$adjust$1$1.invoke(CounterRepository.kt:56)
	at io.realm.internal.SuspendableWriter$write$2.invokeSuspend(SuspendableWriter.kt:67)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -19,6 +19,7 @@
 
 ### Fixed
 * Fixed null pointer exceptions when returning an unmanaged object from `MutableRealm.write/writeBlocking`.
+* Fixed premature closing of underlying realm of frozen objects returned from `MutableRealm.write/writeBlocking`. (Issue [#477](https://github.com/realm/realm-kotlin/issues/477))
 
 ### Compatibility
 * This release is compatible with Kotlin 1.5.21 and Coroutines 1.5.0.

--- a/packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmImpl.kt
+++ b/packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmImpl.kt
@@ -87,12 +87,12 @@ internal class RealmImpl private constructor(configuration: RealmConfigurationIm
 
     override suspend fun <R> write(block: MutableRealm.() -> R): R {
         try {
-            val (nativePointer, versionId, result) = this.writer.write(block)
+            val (reference, result) = this.writer.write(block)
             // Update the user facing Realm before returning the result.
             // That way, querying the Realm right after the `write` completes will return
             // the written data. Otherwise, we would have to wait for the Notifier thread
             // to detect it and update the user Realm.
-            updateRealmPointer(RealmReference(this, nativePointer))
+            updateRealmPointer(reference)
             return result
         } catch (exception: RealmCoreException) {
             throw genericRealmCoreExceptionHandler(

--- a/packages/library-base/src/commonMain/kotlin/io/realm/internal/SuspendableWriter.kt
+++ b/packages/library-base/src/commonMain/kotlin/io/realm/internal/SuspendableWriter.kt
@@ -82,8 +82,8 @@ internal class SuspendableWriter(private val owner: RealmImpl, val dispatcher: C
             // TODO Can we guarantee the Dispatcher is single-threaded? Or otherwise
             //  lock this code?
             val newDbPointer = RealmInterop.realm_freeze(realm.realmReference.dbPointer)
-            // FIXME Should we actually rather just throw if we cannot freeze the result?
             val newReference = RealmReference(owner, newDbPointer)
+            // FIXME Should we actually rather just throw if we cannot freeze the result?
             if (shouldFreezeWriteReturnValue(result)) {
                 result = freezeWriteReturnValue(newReference, result)
             }

--- a/packages/library-base/src/commonMain/kotlin/io/realm/internal/SuspendableWriter.kt
+++ b/packages/library-base/src/commonMain/kotlin/io/realm/internal/SuspendableWriter.kt
@@ -18,8 +18,6 @@ package io.realm.internal
 
 import io.realm.MutableRealm
 import io.realm.RealmObject
-import io.realm.VersionId
-import io.realm.internal.interop.NativePointer
 import io.realm.internal.interop.RealmInterop
 import io.realm.internal.platform.runBlocking
 import io.realm.internal.platform.threadId
@@ -55,7 +53,7 @@ internal class SuspendableWriter(private val owner: RealmImpl, val dispatcher: C
         tid = runBlocking(dispatcher) { threadId() }
     }
 
-    suspend fun <R> write(block: MutableRealm.() -> R): Triple<NativePointer, VersionId, R> {
+    suspend fun <R> write(block: MutableRealm.() -> R): Pair<RealmReference, R> {
         // TODO Would we be able to offer a per write error handler by adding a CoroutineExceptionHandler
         return withContext(dispatcher) {
             var result: R
@@ -84,27 +82,22 @@ internal class SuspendableWriter(private val owner: RealmImpl, val dispatcher: C
             // TODO Can we guarantee the Dispatcher is single-threaded? Or otherwise
             //  lock this code?
             val newDbPointer = RealmInterop.realm_freeze(realm.realmReference.dbPointer)
-            val newVersion = VersionId(RealmInterop.realm_get_version_id(newDbPointer))
             // FIXME Should we actually rather just throw if we cannot freeze the result?
+            val newReference = RealmReference(owner, newDbPointer)
             if (shouldFreezeWriteReturnValue(result)) {
-                result = freezeWriteReturnValue(result, newDbPointer)
+                result = freezeWriteReturnValue(newReference, result)
             }
-            Triple(newDbPointer, newVersion, result)
+            Pair(newReference, result)
         }
     }
 
-    private fun <R> freezeWriteReturnValue(result: R, frozenDbPointer: NativePointer): R {
+    private fun <R> freezeWriteReturnValue(reference: RealmReference, result: R): R {
         return when (result) {
             // is RealmResults<*> -> result.freeze(this) as R
             is RealmObject -> {
                 // FIXME If we could transfer ownership (the owning Realm) in Realm instead then we
                 //  could completely eliminate the need for the external owner in here!?
-                (result as RealmObjectInternal).freezeTyped(
-                    RealmReference(
-                        owner,
-                        frozenDbPointer
-                    )
-                )
+                (result as RealmObjectInternal).freezeTyped(reference)
             }
             else -> throw IllegalArgumentException("Did not recognize type to be frozen: $result")
         }

--- a/test/base/src/androidTest/kotlin/io/realm/test/shared/MutableRealmTests.kt
+++ b/test/base/src/androidTest/kotlin/io/realm/test/shared/MutableRealmTests.kt
@@ -23,6 +23,8 @@ import io.realm.entities.link.Parent
 import io.realm.objects
 import io.realm.test.platform.PlatformUtils
 import io.realm.test.util.Utils.createRandomString
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.runBlocking
 import kotlin.test.AfterTest
 import kotlin.test.BeforeTest
 import kotlin.test.Test
@@ -144,6 +146,21 @@ class MutableRealmTests {
         realm.writeBlocking {
             assertFailsWith<IllegalArgumentException> {
                 val latest = findLatest(StringPropertyWithPrimaryKey())
+            }
+        }
+    }
+
+    @Test
+    fun findLatest_inLongHistory() {
+        runBlocking {
+            val child = realm.write { copyToRealm(Child()) }
+            for (i in 1..10) {
+                realm.write {
+                    findLatest(child)?.apply {
+                        this.name = i.toString()
+                    }
+                }
+                delay(100)
             }
         }
     }

--- a/test/base/src/androidTest/kotlin/io/realm/test/shared/MutableRealmTests.kt
+++ b/test/base/src/androidTest/kotlin/io/realm/test/shared/MutableRealmTests.kt
@@ -162,9 +162,7 @@ class MutableRealmTests {
             val child = realm.write { copyToRealm(Child()) }
             for (i in 1..10) {
                 realm.write {
-                    findLatest(child)?.apply {
-                        this.name = i.toString()
-                    }
+                    assertNotNull(findLatest(child))
                 }
                 delay(100)
             }

--- a/test/base/src/androidTest/kotlin/io/realm/test/shared/RealmTests.kt
+++ b/test/base/src/androidTest/kotlin/io/realm/test/shared/RealmTests.kt
@@ -398,7 +398,7 @@ class RealmTests {
     }
 
     @Test
-    fun closingIntermediateVersionsWhenNoLongerReferenced() {
+    fun intermediateVersionsReleaseWhenProgressingRealm() {
         assertEquals(0, intermediateReferences.value.size)
         realm.writeBlocking { }
         assertEquals(1, intermediateReferences.value.size)
@@ -424,7 +424,7 @@ class RealmTests {
     //  reference on Native. Could just be that the GC somehow does not collect this when
     //  cleared due some thresholds or outcome of GC not being predictable.
     @Ignore
-    fun closingIntermediateVersionsWhenNoLongerReferencedByLocalObject() {
+    fun clearingRealmObjectReleasesRealmReference() {
         assertEquals(0, intermediateReferences.value.size)
         // The below code creates the object without returning it from write to show that the
         // issue is not bound to the freezing inside write, but also happens on the same thread as
@@ -437,9 +437,6 @@ class RealmTests {
         realm.writeBlocking { }
         assertEquals(3, intermediateReferences.value.size)
 
-        // Clear reference
-        parent = null
-
         // Trigger GC - On native we also need to trigger GC on the background thread that creates
         // the references
         runBlocking((realm.configuration as RealmConfigurationImpl).writeDispatcher) {
@@ -449,6 +446,20 @@ class RealmTests {
 
         // Close of intermediate version is currently only done when updating the realm after a write
         realm.writeBlocking { }
+        // We still have the single intermediate reference as a result of the write itself
+        // and the reference kept alive by the realm object
+        assertEquals(2, intermediateReferences.value.size)
+
+        // Clear reference
+        parent = null
+
+        runBlocking((realm.configuration as RealmConfigurationImpl).writeDispatcher) {
+            triggerGC()
+        }
+        triggerGC()
+
+        realm.writeBlocking { }
+        // Clearing the realm object reference allowed clearing the corresponding reference
         assertEquals(1, intermediateReferences.value.size)
     }
 

--- a/test/base/src/androidTest/kotlin/io/realm/test/shared/RealmTests.kt
+++ b/test/base/src/androidTest/kotlin/io/realm/test/shared/RealmTests.kt
@@ -409,7 +409,7 @@ class RealmTests {
 
         // Trigger GC - On native we also need to trigger GC on the background thread that creates
         // the references
-        runBlocking((realm.configuration as RealmConfigurationImpl).writeDispatcher){
+        runBlocking((realm.configuration as RealmConfigurationImpl).writeDispatcher) {
             triggerGC()
         }
         triggerGC()
@@ -420,15 +420,16 @@ class RealmTests {
     }
 
     @Test
-    @Ignore // Clearing local object variable does not trigger collection of reference
-    fun closingIntermediateVersionsWhenNoLongerReferencedByRemoteObject() {
+    // TODO Investigate why clearing local object variable does not trigger collection of
+    //  reference on Native. Could just be that the GC somehow does not collect this when
+    //  cleared due some thresholds or outcome of GC not being predictable.
+    @Ignore
+    fun closingIntermediateVersionsWhenNoLongerReferencedByLocalObject() {
         assertEquals(0, intermediateReferences.value.size)
-        // TODO For some reason cleaning up doesn't work if we have a local reference and clears it.
-        //  The below code creates the object without returning it from write to show that the
-        //  issue is not bound to the freezing inside write, but also happens in a single thread.
-        //  Could just be that the GC somehow does not collect this when cleared due some
-        //  thresholds.
-        realm.writeBlocking { copyToRealm(Parent()); Unit}
+        // The below code creates the object without returning it from write to show that the
+        // issue is not bound to the freezing inside write, but also happens on the same thread as
+        // the realm is constructed on.
+        realm.writeBlocking { copyToRealm(Parent()); Unit }
         var parent: Parent? = realm.objects<Parent>()!!.first()
         assertEquals(1, intermediateReferences.value.size)
         realm.writeBlocking { }
@@ -441,7 +442,7 @@ class RealmTests {
 
         // Trigger GC - On native we also need to trigger GC on the background thread that creates
         // the references
-        runBlocking((realm.configuration as RealmConfigurationImpl).writeDispatcher){
+        runBlocking((realm.configuration as RealmConfigurationImpl).writeDispatcher) {
             triggerGC()
         }
         triggerGC()

--- a/test/base/src/androidTest/kotlin/io/realm/test/shared/RealmTests.kt
+++ b/test/base/src/androidTest/kotlin/io/realm/test/shared/RealmTests.kt
@@ -423,6 +423,7 @@ class RealmTests {
     // TODO Investigate why clearing local object variable does not trigger collection of
     //  reference on Native. Could just be that the GC somehow does not collect this when
     //  cleared due some thresholds or outcome of GC not being predictable.
+    //  https://github.com/realm/realm-kotlin/issues/486
     @Ignore
     fun clearingRealmObjectReleasesRealmReference() {
         assertEquals(0, intermediateReferences.value.size)

--- a/test/base/src/androidTest/kotlin/io/realm/test/shared/RealmTests.kt
+++ b/test/base/src/androidTest/kotlin/io/realm/test/shared/RealmTests.kt
@@ -20,6 +20,7 @@ import io.realm.RealmConfiguration
 import io.realm.VersionId
 import io.realm.entities.link.Child
 import io.realm.entities.link.Parent
+import io.realm.internal.RealmConfigurationImpl
 import io.realm.internal.interop.NativePointer
 import io.realm.internal.platform.WeakReference
 import io.realm.isManaged
@@ -399,15 +400,52 @@ class RealmTests {
     @Test
     fun closingIntermediateVersionsWhenNoLongerReferenced() {
         assertEquals(0, intermediateReferences.value.size)
-        var parent: Parent? = realm.writeBlocking { copyToRealm(Parent()) }
+        realm.writeBlocking { }
         assertEquals(1, intermediateReferences.value.size)
         realm.writeBlocking { }
         assertEquals(2, intermediateReferences.value.size)
+        realm.writeBlocking { }
+        assertEquals(3, intermediateReferences.value.size)
+
+        // Trigger GC - On native we also need to trigger GC on the background thread that creates
+        // the references
+        runBlocking((realm.configuration as RealmConfigurationImpl).writeDispatcher){
+            triggerGC()
+        }
+        triggerGC()
+
+        // Close of intermediate version is currently only done when updating the realm after a write
+        realm.writeBlocking { }
+        assertEquals(1, intermediateReferences.value.size)
+    }
+
+    @Test
+    @Ignore // Clearing local object variable does not trigger collection of reference
+    fun closingIntermediateVersionsWhenNoLongerReferencedByRemoteObject() {
+        assertEquals(0, intermediateReferences.value.size)
+        // TODO For some reason cleaning up doesn't work if we have a local reference and clears it.
+        //  The below code creates the object without returning it from write to show that the
+        //  issue is not bound to the freezing inside write, but also happens in a single thread.
+        //  Could just be that the GC somehow does not collect this when cleared due some
+        //  thresholds.
+        realm.writeBlocking { copyToRealm(Parent()); Unit}
+        var parent: Parent? = realm.objects<Parent>()!!.first()
+        assertEquals(1, intermediateReferences.value.size)
+        realm.writeBlocking { }
+        assertEquals(2, intermediateReferences.value.size)
+        realm.writeBlocking { }
+        assertEquals(3, intermediateReferences.value.size)
 
         // Clear reference
         parent = null
-        // Trigger GC
+
+        // Trigger GC - On native we also need to trigger GC on the background thread that creates
+        // the references
+        runBlocking((realm.configuration as RealmConfigurationImpl).writeDispatcher){
+            triggerGC()
+        }
         triggerGC()
+
         // Close of intermediate version is currently only done when updating the realm after a write
         realm.writeBlocking { }
         assertEquals(1, intermediateReferences.value.size)

